doc/developer/design/20231027_refresh_mvs.md

```diff
@@ -0,0 +1,279 @@
+
+### Implementation Alternative: Suspending the Replica Process to Disk
+
+As mentioned in the [scoping section](#out-of-scope), an alternative implementation that would cover also intermediate refresh intervals would be to suspend the replica process and save its entire state to S3 and a local disk cache, and restore it when a refresh is needed. [There are existing techniques for saving process state, e.g., CRIU.](https://faun.pub/kubernetes-checkpointing-a-definitive-guide-33dd1a0310f6) However, this would be a much bigger project than the proposed solution.
+
+## Automated Cluster Scheduling for `REFRESH EVERY` MVs
+
+TODO: updated discussion
+
+## Introspection / Observability
+
+For showing the `REFRESH` options of each materialized view, I'm planning to create a new table in `mz_internal` called `mz_materialized_view_refresh_options`, which will have a row for each non-trivial refresh option of each MV. The columns will be `(materialized_view_id text, refresh_type text, refresh_interval interval, refresh_interval_aligned_to timestamptz, refresh_at timestamptz)`. In `refresh_type` we would show either the string "REFRESH EVERY" or "REFRESH AT". From those columns that show details of a particular refresh option, only those columns will have a non-null value that are relevant for the refresh option that the row is about: a `REFRESH EVERY` would be non-null in `interval` and `interval_aligned_to`, a `REFRESH AT` would be non-null in `refresh_at`. `REFRESH ON COMMIT` is considered a trivial refresh option, and won't show up in the table. (See Nikhil's [msg](https://materializeinc.slack.com/archives/C06535JL58R/p1701228087745239?thread_ts=1700595291.424189&cid=C06535JL58R) on an old design thread.)
+
+For showing the previous and next refresh of each MV, I'm planning to create a new table in `mz_internal` called `mz_materialized_view_refreshes (materialized_view_id text, last_completed_refresh timestamptz, next_refresh timestamptz)`. I'm thinking to update it in `Controller::record_frontiers` (where `mz_frontiers` is also updated). The values will be calculated as follows:
+- `next_refresh` will usually be the write frontier of the MV's storage collection, unless:
+  - The write frontier is 0. This happens when the first refresh hasn't completed yet, in which case we need to figure out the first refresh: we'll call `RefreshSchedule::round_up_timestamp` on the MV's `as_of`.
```

Comment:
I'd like to reserve the right to initialize the write frontier to `as_of` as well, so I suggest changing the edge case to "the write frontier is before or at the `as_of`". I think the solution is the same though: logging the first refresh as the `next_refresh` at collection creation.

Reply:
Done